src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/CSharpConvertToRecordRefactoringProvider.cs

```diff
@@ -0,0 +1,791 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertToRecord
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.ConvertToRecord), Shared]
+    internal sealed class CSharpConvertToRecordRefactoringProvider : CodeRefactoringProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpConvertToRecordRefactoringProvider()
+        {
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, span, cancellationToken) = context;
+
+            var typeDeclaration = await context.TryGetRelevantNodeAsync<TypeDeclarationSyntax>().ConfigureAwait(false);
+            if (typeDeclaration == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            if (semanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken) is not INamedTypeSymbol
+                {
+                    // if type is an interface we don't want to refactor
+                    TypeKind: TypeKind.Class or TypeKind.Struct,
+                    // no need to convert, already a record
+                    // TODO: when adding other options it might make sense to convert a postional record to a non-positional one and vice versa
+                    IsRecord: false,
+                    // records can't be static and so if the class is static we probably shouldn't convert it
+                    IsStatic: false
+
+                } type ||
+                // make sure that there is at least one positional parameter we can introduce
+                !typeDeclaration.Members.Any(m => ShouldConvertProperty(m, type)))
+            {
+                return;
+            }
+
+            var positionalTitle = CSharpFeaturesResources.Convert_to_positional_record;
+
+            var positional = CodeAction.Create(
+                positionalTitle,
+                cancellationToken => ConvertToParameterRecordAsync(document, type, typeDeclaration, cancellationToken),
+                positionalTitle);
```

Comment:
```suggestion
                nameof(CSharpFeaturesResources.Convert_to_positional_record));
```

(this defends us against the unlikely, but something which has happened, where there are different equivalence keys, but they get localized to the same title).